### ISSUE TITLE
Rule Suggestion: Throw an error on reassignment of variables that sit on globals

### ISSUE THREAD

**Issue by sscaff1** [eslint/eslint#10857]
**Please describe what the rule should do:**
The rule should throw an error if you attempt to destructure or reassign variables on configured variables. 
**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**
Assume the eslint rule throws an error for the DOM globals `window` and `document`. Then the following would throw errors.

```js
  const { myVar } = window;
  const myVar = window.myVar;
  const { body } = document;
  const body = window.body;
```
Currently, I am trying to modularize a legacy code base that relied heavily on globals. Because variables can be reassigned from the window object, it's very difficult to look through a code base with tens of thousands of lines and find all references coming from the window object. We get into these situations too which makes this even harder:
```js
const { myVar } = window;
// 100 lines later
const { myNestedVar } = myVar;
```
Any suggestions are appreciated but this could really help us out to ensure developers don't create more references like this during the modularization process (which isn't a short term project). 

**Why should this rule be included in ESLint (instead of a plugin)?**
This seems similar to http://docs.w3cub.com/eslint/rules/no-global-assign/ (at least the rule would be written in the config in much the same way). I guess it could go either way but this doesn't really have any dependencies

**Comment by platinumazure** [comment]
Hi @sscaff1, thanks for the issue.

It seems that the [`no-restricted-globals`](https://eslint.org/docs/rules/no-restricted-globals) could help in your case. You could configure it to flag all references to `window` and `document`. Give it a try and let us know if it doesn't help you.

**Comment by platinumazure** [comment]
Thanks @sscaff1. I apologize, I might have misunderstood what you were trying to do.

I was going by this statement of yours from the initial post:

> Currently, I am trying to modularize a legacy code base that relied heavily on globals. Because variables can be reassigned from the window object, it's very difficult to look through a code base with tens of thousands of lines and find all references coming from the window object.

The approach I've suggested will allow you to proceed with the following steps:

1. Enable `no-restricted-globals` for window, document, and any other globals you want to target. Also enable `no-undef` if it's not already enabled.
1. Run ESLint and find the global references (i.e., look for `no-restricted-globals` errors).
1. Edit the file to remove any destructuring or other property aliasing assignments.
1. Run ESLint on the file again. Find and fix the `no-undef` errors, which should arise from removing the declaration of variables like `body`.
    a. If needed, repeat this process for any later redeclarations of the destructured properties.

It's a little tedious, but it's much faster than trying to read the files yourself.

Tracking destructuring and reassignments in an ESLint rule, on the other hand, could prove cumbersome. And I'm not (yet) convinced that we should add or modify a core ESLint rule to support this use case, when it seems that there is another approach that could be used.

Hope this helps! Please let me know if I'm still misunderstanding something.

**Comment by sscaff1** [comment]
This is great! A little tedious but it'll work. Thanks for that. I'll close this issue.
